components/locid/src/extensions/unicode/mod.rs

```diff
@@ -151,87 +155,57 @@ impl Unicode {
-        let mut attributes = ShortBoxSlice::new();
-
-        while let Some(subtag) = iter.peek() {
-            if let Ok(attr) = Attribute::try_from_bytes(subtag) {
-                if let Err(idx) = attributes.binary_search(&attr) {
-                    attributes.insert(idx, attr);
-                }
-            } else {
-                break;
-            }
-            iter.next();
-        }
-
-        let mut keywords = LiteMap::new();
-
-        let mut current_keyword = None;
-        let mut current_value = ShortBoxSlice::new();
-
-        while let Some(subtag) = iter.peek() {
-            let slen = subtag.len();
-            if slen == 2 {
-                if let Some(kw) = current_keyword.take() {
-                    keywords.try_insert(kw, Value::from_short_slice_unchecked(current_value));
-                    current_value = ShortBoxSlice::new();
-                }
-                current_keyword = Some(Key::try_from_bytes(subtag)?);
-            } else if current_keyword.is_some() {
-                match Value::parse_subtag(subtag) {
-                    Ok(Some(t)) => current_value.push(t),
-                    Ok(None) => {}
-                    Err(_) => break,
-                }
-            } else {
-                break;
-            }
-            iter.next();
-        }
-
-        if let Some(kw) = current_keyword.take() {
-            keywords.try_insert(kw, Value::from_short_slice_unchecked(current_value));
-        }
+        let attributes = Attributes::try_from_iter(iter)?;
+        let keywords = Keywords::try_from_iter(iter)?;
 
         // Ensure we've defined at least one attribute or keyword
         if attributes.is_empty() && keywords.is_empty() {
             return Err(ParserError::InvalidExtension);
         }
 
         Ok(Self {
-            keywords: keywords.into(),
-            attributes: Attributes::from_short_slice_unchecked(attributes),
+            keywords,
+            attributes,
         })
     }
 
-    pub(crate) fn for_each_subtag_str<E, F>(&self, f: &mut F) -> Result<(), E>
+    pub(crate) fn for_each_subtag_str<E, F>(&self, f: &mut F, with_ext: bool) -> Result<(), E>
     where
         F: FnMut(&str) -> Result<(), E>,
     {
-        if self.is_empty() {
-            return Ok(());
+        if !self.is_empty() {
+            if with_ext {
+                f("u")?;
+            }
+            self.attributes.for_each_subtag_str(f)?;
+            self.keywords.for_each_subtag_str(f)?;
         }
-        f("u")?;
-        self.attributes.for_each_subtag_str(f)?;
-        self.keywords.for_each_subtag_str(f)?;
         Ok(())
     }
 }
 
+impl FromStr for Unicode {
+    type Err = ParserError;
+
+    fn from_str(source: &str) -> Result<Self, Self::Err> {
+        Self::try_from_bytes(source.as_bytes())
+    }
+}
+
 writeable::impl_display_with_writeable!(Unicode);
 
 impl writeable::Writeable for Unicode {
```

Comment:
It seems somewhat risky to change the `Writeable` to not include the `u-` as it has been doing since 1.0. It's easy for me to imagine code that depended on that behavior.

Reply:
I think we haven't thought it through sufficiently and I consider 2.0 a chance to normalize that. Stringification of a struct doesn't contain the extension key of the struct.

Reply:
Ok, we can start landing 2.0 changes now. However, I'm not fully convinced. In the case of Other especially, it seems like printing out the extension key is very much the job of the extension struct. But, if you want to make this change (impl Writeable does not include `u`), please do so consistently across Private, Other, and Transform.

Reply:
That's fair. I'm convinced by the compatibility with Other. I normalized the behavior across all extension types to require extension key for parsing and print it in response